### ISSUE TITLE
Inefficient build

### ISSUE THREAD

**Issue by Jack-McDowell** [matthewbdwyer/tipc#22]
The TIPC build compiles static libraries, but it doesn't seem to properly use them for some reason. When a file is modified, it is often recompiled up to 7 times per build depending on what part of the project it's in. This results in significantly increased build times.

The following screenshot illustrates the issue as it manifests with modifications to the type unifier.
![image](https://user-images.githubusercontent.com/10685137/116165433-e2e18580-a6c9-11eb-9627-ffd83f6652ee.png)

The unifier is recompiled to a .o file seven times, once per artifact. Ideally, it should be compiled once into a static library, which should just be linked when needed rather than recompiling each time.

Unfortunately, I don't know CMake well enough to point out the specific issue in the build process leading to this behavior.

**Comment by nicholasphair** [comment]
Hi Jack, good find. Can you share more details about this build? eg. build command, build directory, platform, cmake version. Anything relevant to reproducing. It wouldn't shock me if the CMake setup leads to this kind of behavior. I'm certainly not a CMake expert either :). 

One note is that it does look like those object files are different - the fqpns are not the same. Of course, they probably should be.

**Comment by Jack-McDowell** [comment]
I'm on LLVM 11.1 and CMake 3.16.3 running in WSL on Ubuntu 20.02

The full path to my build directory is `/mnt/c/Users/Jack/source/repos/tipc/build`. This was created by cloning the tipc repository and running `mkdir build` inside the tipc directory. Inside the build folder, I ran `cmake ..` to prepare the cmake files, then `make -j8` to compile the whole project. 

After setting it up, I made a change to unifier.cpp, which resulted in this behavior when I rebuilt with `make -j8`.

The object files may be different, but it's the same source file being compiled multiple times. And it looks like there are four different object files being created; Unifier.cpp.o is being written to test/unit four times.
